packages/tldraw/src/lib/ui/components/DebugPanel.tsx

```diff
@@ -68,6 +82,39 @@ const CurrentState = track(function CurrentState() {
 	return <div className="tlui-debug-panel__current-state">{editor.getPath()}</div>
 })
 
+const times: number[] = []
+let isFPSCounterRunning = false
+let fps = 0
+function refreshFpsLoop() {
+	window.requestAnimationFrame(function () {
+		const now = performance.now()
+		while (times.length > 0 && times[0] <= now - 1000) {
+			times.shift()
+		}
+		times.push(now)
+		fps = times.length
+		isFPSCounterRunning && refreshFpsLoop()
+	})
+}
+
+function FPS() {
+	const [rerender, setRerender] = React.useState(0)
+	React.useEffect(() => {
+		setTimeout(() => setRerender(rerender + 1), 33)
+	}, [rerender])
+
+	const isLow = fps < 90
+	return (
+		<div
+			className={classNames('tlui-debug-panel__fps', {
+				'tlui-debug-panel__fps__slow ': isLow,
+			})}
+		>
+			FPS {fps}
+		</div>
+	)
+}
```

Comment:
My instinct here would be to put the frame-by-frame stuff outside of the react lifecycle (i.e. with a little canvas). We'd be adding noise to the system by pushing frame-to-frame updates through React.

Reply:
ie the test should be as separate as possible from the system it is measuring

Reply:
talked about this offline, just repeating here for context:
but yeah, totally. right now `refreshFpsLoop` is measuring outside of the React lifecycle and this render loop inside the component is just to "grab" that measurement number from time-to-time. We could make it grab it once a second if we wanted to it instead of every 33ms

Reply:
A lot of our performance comes from how (and how often) the React tree updates. By pushing a state change through React every frame, we are adding stress to that system and will add noise to other debugging tools. My suggestion is to avoid react state when updating the count: i.e. run the loop inside of an effect and imperatively paint a canvas or update an element's innerText, just to avoid that per-frame react state update.